### PR TITLE
Add an Euler model

### PR DESCRIPTION
**Pull request by thomasgibson**
This PR adds a simple Euler model using a nodal DG discretization. This operator is the basis for the corresponding ESDG Euler operator in the entropy stable development branches and will be updated in a future PR related to flux differencing.

### CODE REVIEW

File: examples/euler/acoustic_pulse.py [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+import logging
+logger = logging.getLogger(__name__)
+
+
+def rk4_step(y, t, h, f):
+    k1 = f(t, y)
+    k2 = f(t+h/2, y + h/2*k1)
+    k3 = f(t+h/2, y + h/2*k2)
+    k4 = f(t+h, y + h*k3)
+    return y + h/6*(k1 + 2*k2 + 2*k3 + k4)
```
Comment (inducer) [inline examples/euler/acoustic_pulse.py:59]: Not a huge fan of repeating this all over. Maybe put it someplace central and import it?

File: grudge/models/euler.py [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+
+# {{{ Array container for the Euler model
+
+@with_container_arithmetic(bcast_obj_array=False,
+                           bcast_container_types=(DOFArray, np.ndarray),
+                           matmul=True,
+                           rel_comparison=True)
+@dataclass_array_container
+@dataclass(frozen=True)
+class EulerContainer:
```
Comment (inducer) [inline grudge/models/euler.py:54]:
```suggestion
class EulerField:
```
?

File: examples/euler/acoustic_pulse.py [hunk 3]
```diff
@@ -0,0 +1,271 @@
+"""Minimal example of a grudge driver."""
```
Comment (inducer) [inline examples/euler/acoustic_pulse.py:1]: Delete or improve. (also below)

File: grudge/models/euler.py [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+import numpy as np
```
Comment (inducer) [inline grudge/models/euler.py:25]: I'd like for this to be backed by convergence tests. Maybe the vortex could be moved to someplace importable and used in this manner?

**Review by inducer — COMMENTED** [review]
Thanks! A few thoughts from a preliminary look below.

File: examples/euler/acoustic_pulse.py (hunk 1 above)
Comment (thomasgibson) [reply examples/euler/acoustic_pulse.py]: Done! https://github.com/inducer/grudge/pull/196/commits/97a8e47d2da1f0916c13e0cb3b626dfd9265e44e

File: examples/euler/acoustic_pulse.py (hunk 3 above)
Comment (thomasgibson) [reply examples/euler/acoustic_pulse.py]: Done: https://github.com/inducer/grudge/pull/196/commits/ee72ab20c85a96df50f4b6e48c75e5a43c1a5b75

File: grudge/models/euler.py (hunk 2 above)
Comment (thomasgibson) [reply grudge/models/euler.py]: Sure, that sounds much better: https://github.com/inducer/grudge/pull/196/commits/f5e0b62a67be603ff109b98a13883ba591858521

File: grudge/models/euler.py (hunk 4 above)
Comment (thomasgibson) [reply grudge/models/euler.py:25]: Good idea --- I've gone ahead and set up a convergence test: https://github.com/inducer/grudge/pull/196/commits/c10097d04870ba35a33a5a90db73b83045057eea

File: grudge/models/euler.py [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+
+# }}}
+
+
+# {{{ Predefined initial conditions for the Euler model
+
+def vortex_initial_condition(x_vec, t=0):
+    """Initial condition adapted from Section 2 (equation 2) of:
+
+    - K. Mattsson, M. Sv\"{a}rd, M. Carpenter, and J. Nordstr\"{o}m (2006).
```
Comment (inducer) [inline grudge/models/euler.py:77]:
```suggestion
    - K. Mattsson, M. Svärd, M. Carpenter, and J. Nordström (2006).
```

File: grudge/models/euler.py [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+    High-order accurate computations for unsteady aerodynamics.
+    [DOI](https://doi.org/10.1016/j.compfluid.2006.02.004).
```
Comment (inducer) [inline grudge/models/euler.py:79]:
```suggestion
    `DOI <https://doi.org/10.1016/j.compfluid.2006.02.004>`__.
```
reST is weird, I know.

File: grudge/models/euler.py [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+    """
+    mach = 0.5    # Mach number
+    _x0 = 5
+    epsilon = 1   # vortex strength
+    gamma = 1.4
```
Comment (inducer) [inline grudge/models/euler.py:84]: Make these (default) arguments?

File: grudge/models/euler.py [hunk 8]
```diff
@@ -0,0 +1,364 @@
...
+
+# {{{ Array container for the Euler model
+
+@with_container_arithmetic(bcast_obj_array=False,
+                           bcast_container_types=(DOFArray, np.ndarray),
+                           matmul=True,
+                           rel_comparison=True)
+@dataclass_array_container
+@dataclass(frozen=True)
+class EulerField:
```
Comment (inducer) [inline grudge/models/euler.py:56]:
```suggestion
class ConservedEulerField:
```

File: grudge/models/euler.py [hunk 9]
```diff
@@ -0,0 +1,364 @@
...
+    :returns: A :class:`Tuple` containing the primitive variables:
+        (density, velocity, pressure).
+    """
+    rho = cv_state.mass
+    rho_e = cv_state.energy
+    rho_u = cv_state.momentum
+    u = rho_u / rho
+    p = (gamma - 1) * (rho_e - 0.5 * sum(rho_u * u))
+
+    return rho, u, p
```
Comment (inducer) [inline grudge/models/euler.py:124]: Have a dataclass for these, too?

File: grudge/models/euler.py [hunk 10]
```diff
@@ -0,0 +1,364 @@
+"""Grudge operators modeling compressible, inviscid flows (Euler)"""
```
Comment (inducer) [inline grudge/models/euler.py:1]: Hook this into the docs.

Comment (inducer) [inline grudge/models/euler.py:1]: Sprinkle some type annotations? Doesn't have to be comprehensive.

**Review by inducer — APPROVED** [review]
Thanks for working on this! A few minor things, then this is good to go from my end.

File: grudge/models/euler.py (hunk 6 above)
Comment (thomasgibson) [reply grudge/models/euler.py]: Done! https://github.com/inducer/grudge/pull/196/commits/79eaac664a85a0d407a7e63c8042b6178f9d2d8e

File: grudge/models/euler.py (hunk 5 above)
Comment (thomasgibson) [reply grudge/models/euler.py]: Done! https://github.com/inducer/grudge/pull/196/commits/79eaac664a85a0d407a7e63c8042b6178f9d2d8e

File: grudge/models/euler.py (hunk 8 above)
Comment (thomasgibson) [reply grudge/models/euler.py]: Renamed: https://github.com/inducer/grudge/pull/196/commits/2f12b1edb331bdef86e76dfb45b30208ec4086ba

File: grudge/models/euler.py (hunk 10 above)
Comment (thomasgibson) [reply grudge/models/euler.py:1]: Sprinkled in https://github.com/inducer/grudge/pull/196/commits/2f12b1edb331bdef86e76dfb45b30208ec4086ba

File: grudge/models/euler.py (hunk 9 above)
Comment (thomasgibson) [reply grudge/models/euler.py:123]: Ehh I think that might be a bit excessive for these. They're only really used internally (except if explicitly requested via the documented helper routine `conservative_to_primitive_vars` for plotting purposes).

Comment (inducer) [reply grudge/models/euler.py:145]: Fair enough. Also easy to hack in later.

**Pull request merged**